c10::string_view

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26628 [wip] Remove CAFFE_KNOWN_TYPE
* #26619 [wip] caffe2::TypeMeta uses compile time type names
* #26618 [wip] Compile time type names
* **#26616 c10::string_view**

Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrieval which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

smessmer committed:
[wip] c10::util::string_view

Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

ezyang commented on c10/util/string_view.h:
```diff
@@ -0,0 +1,677 @@
+namespace c10 {
+
+/**
+ * Reimplementation of std::string_view for C++11.
```
Umm... did you implement this yourself?

smessmer replied:
yes, I didn't find a public implementation that fit our needs i.e. is constexpr and works with pre-C++17.

ezyang commented:
> retrievaly

What's retrievaly?

ezyang commented:
I'm not sure I have time to do a thorough review of a reimplementation of `string_view` from scratch...

smessmer committed:
Update on "c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrievaly which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer commented:
This is quite well tested, I probably don't need too thoroughly a review if you don't have the time.

> > retrievaly
> 
> What's retrievaly?

retrieval

smessmer committed:
Update on "c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrieval which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrieval which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrieval which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrieval which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrieval which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrieval which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrieval which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrieval which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrieval which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrieval which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrieval which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

smessmer committed:
Update on "c10::string_view"


Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrieval which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

[ghstack-poisoned]

dzhulgakov reviewed (approved):
Ok, I read it diagonally and looks plausible. Given extensive testing and that it's mostly constexpr it should be good

gchanan commented on c10/util/C++17.h:
```diff
@@ -294,6 +294,15 @@ template<class T> inline std::string to_string(T value) {
 }
 
+template <class T>
+constexpr const T& min(const T& a, const T& b) {
```
this is constexpr in C++14, right?  Is there a reason we aren't tracking C++14 / C++17 separately so we can kill the C++14 stuff once we move to that?
